### PR TITLE
Improve vmap out_axes error

### PR DESCRIPTION
**Pull request by cgarciae**
## Changes

Shows the path of invalid leaves when trying to return vectorized arrays through outputs marked as `out_axes=None`.

### CODE REVIEW

File: jax/_src/api.py [hunk 1]
```diff
@@ -1211,11 +1211,22 @@ def vmap_f(*args, **kwargs):
...
+    except batching.SpecMatchError as e:
+      # TODO this is a bit circular, maybe there's a better way to get paths..
+      # let's look at how shard_map organizes things?
+      out_axes_flat = flatten_axes("vmap out_axes", out_tree(), out_axes)
+      out_axes_full = tree_unflatten(out_tree(), out_axes_flat)
+      pairs, _ = tree_flatten_with_path(out_axes_full, is_leaf=lambda x: x is None)
+
+      path, _ = pairs[e.leaf_idx]
+      raise ValueError(f'at vmap out_axes{keystr(path)}, got axis spec {e.dst} '
+                       f'but output was batched on axis {e.src}')
```
Comment (jakevdp) [inline jax/_src/api.py:1229]: Should we add `from e`?

**Comment by cgarciae** [comment]
@mattjj I fixed the remaining tests. There's still the TODO of trying to refactor the code such that we don't flatten twice but maybe we can do this in the future? This basic implementation is significantly more useful than what we had before.

File: jax/_src/api.py (hunk 1 above)
Comment (cgarciae) [reply jax/_src/api.py:1229]: I think the `batching.SpecMatchError` is a synthetic error we raise just to be able to capture it here to extract its metadata, the user doesn't really doesn't gain any information from it.

Comment (mattjj) [reply jax/_src/api.py:1229]: Yeah I think `from None` may be better here actually!

File: jax/_src/api.py [hunk 2]
```diff
@@ -1211,11 +1211,22 @@ def vmap_f(*args, **kwargs):
...
-    ).call_wrapped(*args_flat)
+    try:
+      out_flat = batching.batch(
+          flat_fun, axis_name, axis_size_, in_axes_flat,
+          lambda: flatten_axes("vmap out_axes", out_tree(), out_axes),
+          spmd_axis_name=spmd_axis_name
+      ).call_wrapped(*args_flat)
+    except batching.SpecMatchError as e:
+      # TODO this is a bit circular, maybe there's a better way to get paths..
+      # let's look at how shard_map organizes things?
```
Comment (mattjj) [inline jax/_src/api.py:1222]:
```suggestion
```

**Review by mattjj — APPROVED** [review]

File: jax/_src/api.py (hunk 1 above)
Comment (mattjj) [inline jax/_src/api.py:1229]:
```suggestion
                       f'but output was batched on axis {e.src}') from None
```

**Pull request merged**